torch/distributed/tensor/parallel/fsdp.py

```diff
@@ -220,6 +221,66 @@ def _chunk_tensor(
         )
 
 
+def _chunk_dtensor(
+    tensor: torch.Tensor,
+    rank: int,
+    device_mesh: DeviceMesh,
+) -> DTensor:
+    """
+    Shard a tensor to chunks along the first dimension. The local rank will gets its
+    corresponding chunk as the local tensor to create a DTensor.
+    """
+    parent_mesh = mesh_resources.get_parent_mesh(device_mesh)
+    if parent_mesh is None:
+        raise RuntimeError("No parent device_mesh is found for FSDP device_mesh.")
+    if parent_mesh.ndim != 2:
+        raise RuntimeError(
+            f"Found parent device_mesh of ndim={parent_mesh.ndim},",
+            "but only 2D meshes are currently supported.",
+        )
```

Comment:
Good point. Maybe we should add a check in TP so that would happen during initialization so that would be earlier than state_dict call. For example, during TP ctor, if TP has a parent mesh and the dim idx of TP is not the innermost dimenstion, we error out say we don't support the composition. Wdyt?

Reply:
Added composition validation logic during TP initialization in https://github.com/pytorch/pytorch/pull/111001